enterprise/server/invocation_search_service/invocation_search_service.go

```diff
@@ -94,63 +99,58 @@ func addPermissionsCheckToQuery(u interfaces.UserInfo, q *query_builder.Query) {
 	o.AddOr(groupQueryStr, groupArgs...)
 	o.AddOr("(i.perms & ? != 0 AND i.user_id = ?)", perms.OWNER_READ, u.GetUserID())
 	orQuery, orArgs := o.Build()
-	q = q.AddWhereClause("("+orQuery+")", orArgs...)
+	q.AddWhereClause("("+orQuery+")", orArgs...)
 }
 
-func (s *InvocationSearchService) QueryInvocations(ctx context.Context, req *inpb.SearchInvocationRequest) (*inpb.SearchInvocationResponse, error) {
-	if err := s.checkPreconditions(req); err != nil {
-		return nil, err
-	}
-	u, err := perms.AuthenticatedUser(ctx, s.env)
-	if err != nil {
-		return nil, err
-	}
-	if blocklist.IsBlockedForStatsQuery(u.GetGroupID()) {
-		return nil, status.ResourceExhaustedErrorf("Too many rows.")
+func containsExecutionColumn(f []*sfpb.StatFilter) bool {
+	for _, filter := range f {
+		if filter.GetMetric().Execution != nil {
+			return true
+		}
 	}
+	return false
+}
 
-	q := query_builder.NewQuery(`SELECT * FROM Invocations as i`)
-
+// This function adds where clauses that can be applied to both invocation- and
+// execution-based searches.  Adding these filters to both phases of an
+// execution-based search is duplicative, but we do it anyway to shrink the
+// result set early while still ensure that our final search in the invocations
+// table matches the query style of "normal" invocations.
+func addCommonWhereClauses(q *query_builder.Query, iq *inpb.InvocationQuery) {
 	// Don't include anonymous builds.
 	q.AddWhereClause("((i.user_id != '' AND i.user_id IS NOT NULL) OR (i.group_id != '' AND i.group_id IS NOT NULL))")
 
-	if user := req.GetQuery().GetUser(); user != "" {
+	if user := iq.GetUser(); user != "" {
 		q.AddWhereClause("i.user = ?", user)
 	}
-	if host := req.GetQuery().GetHost(); host != "" {
+	if host := iq.GetHost(); host != "" {
 		q.AddWhereClause("i.host = ?", host)
 	}
-	if url := req.GetQuery().GetRepoUrl(); url != "" {
+	if url := iq.GetRepoUrl(); url != "" {
 		q.AddWhereClause("i.repo_url = ?", url)
 	}
-	if branch := req.GetQuery().GetBranchName(); branch != "" {
+	if branch := iq.GetBranchName(); branch != "" {
 		q.AddWhereClause("i.branch_name = ?", branch)
 	}
-	if command := req.GetQuery().GetCommand(); command != "" {
+	if command := iq.GetCommand(); command != "" {
 		q.AddWhereClause("i.command = ?", command)
 	}
-	if sha := req.GetQuery().GetCommitSha(); sha != "" {
+	if sha := iq.GetCommitSha(); sha != "" {
 		q.AddWhereClause("i.commit_sha = ?", sha)
 	}
-	if group_id := req.GetQuery().GetGroupId(); group_id != "" {
+	if group_id := iq.GetGroupId(); group_id != "" {
```

Comment:
if we are using clickhouse, i think group_id is required for searching invocation or executions. Without group_id, it will result in full scan.